opa/opa.go

```diff
@@ -0,0 +1,188 @@
+package opa
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/server/types"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/observatorium/observatorium/rbac"
+)
+
+// Input models the data that is used for OPA input documents.
+type Input struct {
+	Groups     []string        `json:"groups"`
+	Permission rbac.Permission `json:"permission"`
+	Resource   string          `json:"resource"`
+	Subject    string          `json:"subject"`
+	Tenant     string          `json:"tenant"`
+}
+
+type config struct {
+	logger     log.Logger
+	registerer prometheus.Registerer
+}
+
+// Option modifies the configuration of an OPA authorizer.
+type Option func(c *config)
+
+// LoggerOption sets a custom logger for the authorizer.
+func LoggerOption(logger log.Logger) Option {
+	return func(c *config) {
+		c.logger = logger
+	}
+}
+
+// RegistererOption sets a Prometheus registerer for the authorizer.
+func RegistererOption(r prometheus.Registerer) Option {
+	return func(c *config) {
+		c.registerer = r
+	}
+}
+
+type restAuthorizer struct {
+	client *http.Client
+	url    *url.URL
+
+	logger     log.Logger
+	registerer prometheus.Registerer
+}
+
+// Authorize implements the rbac.Authorizer interface.
+func (a *restAuthorizer) Authorize(subject string, groups []string, permission rbac.Permission, resource, tenant string) bool {
+	var i interface{} = Input{
+		Groups:     groups,
+		Permission: permission,
+		Resource:   resource,
+		Subject:    subject,
+		Tenant:     tenant,
+	}
+	dreq := types.DataRequestV1{
+		Input: &i,
+	}
+	j, err := json.Marshal(dreq)
+	if err != nil {
+		level.Error(a.logger).Log("msg", "failed to marshal OPA input to JSON", "err", err.Error())
+		return false
+	}
+
+	res, err := a.client.Post(a.url.String(), "application/json", bytes.NewBuffer(j))
+	if err != nil {
+		level.Error(a.logger).Log("msg", "make request to OPA endpoint", "URL", a.url.String(), "err", err.Error())
+		return false
+	}
+
+	if res.StatusCode != 200 {
```

Comment:
Pretty nice is also doing 
```suggestion
	if res.StatusCode/100 != 2 {
```
That way it doesn't matter if 200 or 201 or whatever. :)
It's totally fine like that though.

Reply:
:ok_hand: